File: 0463-island-perimeter/0463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        queue = deque()
        directions = ((-1, 0), (1, 0), (0, -1), (0, 1))
        m, n = len(grid), len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    queue.append((i, j))
        ans = 4 * len(queue)
        while queue:
            x, y = queue.popleft()
            
            for dx, dy in directions:
                if 0 <= x + dx < len(grid) and 0 <= y + dy < len(grid[0]) and grid[x + dx][y + dy] == 1:
                    ans -= 1
                
        return ans
        

File: 0463-island-perimeter/0463-island-perimeter.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        int rows = grid.size(), cols = grid[0].size(), ans = 0;
        
        for (int r = 0; r < rows; ++r){
            for (int c = 0; c < cols; ++c){
                if (grid[r][c] == 1){
                    ans += 4;
                    if (r > 0 && grid[r - 1][c] == 1) ans -= 2;
                    if (c > 0 && grid[r][c - 1] == 1) ans -= 2;
                }
            }
        }
        
        return ans;
    }
};